zerver/lib/timestamp.py

```diff
@@ -42,3 +45,34 @@ def timestamp_to_datetime(timestamp: float) -> datetime.datetime:
 def datetime_to_timestamp(dt: datetime.datetime) -> int:
     verify_UTC(dt)
     return calendar.timegm(dt.timetuple())
+
+def parse_generic_time(timestr: Text) -> Optional[datetime.datetime]:
+    GENERIC_TIMESTR = re.compile('.*\s*(\d+|an|a)\s(\w+)\s*.*', re.I)
+    generic_time = GENERIC_TIMESTR.search(timestr)
+    parsed_time = timezone_now()
+    if generic_time:
+        try:
+            delta = int(generic_time.group(1))
+        except ValueError:
+            delta = 1
+        if delta < 1:
+            return None
+        delta_type = generic_time.group(2).lower()
+        if delta_type in ['min', 'mins', 'minute', 'minutes']:
+            parsed_time += datetime.timedelta(minutes=delta)
+        if delta_type in ['hour', 'hours']:
+            parsed_time += datetime.timedelta(hours=delta)
+        if delta_type in ['day', 'days']:
+            parsed_time += datetime.timedelta(days=delta)
+        if delta_type in ['week', 'weeks']:
+            parsed_time += datetime.timedelta(weeks=delta)
+        return parsed_time
+    if 'next week' == timestr.lower():
+        parsed_time += datetime.timedelta(days=7 - parsed_time.weekday())
+        parsed_time = parsed_time.replace(hour=9, minute=0, second=0, microsecond=0)
+        return parsed_time
+    if 'tomorrow' == timestr.lower():
```

Comment:
I think for implementing tomorrow, I would do: If the input has "tomorrow", replace tomorrow with "March 1" (or whatever the appropriate date is), and then pass the whole string to dateutil.parser.
Same with anything else that can be done that way (e.g. can we make "3am on Tuesday" work using something like this?)

Reply:
I would also just start with lowercasing the timestr at the beginning of the function; doing otherwise I think is just going to open surface area for bugs.

Reply:
agreed, I should have lower cased the string at the beginning only. Replacing is also a good way to do it. I will give that a try.